File: processout-sdk/src/main/java/com/processout/processout_sdk/CustomerAction.java
package com.processout.processout_sdk;

import com.google.gson.annotations.SerializedName;

class CustomerAction {

    protected enum CustomerActionType {
        @SerializedName("fingerprint-mobile")
        FINGERPRINT_MOBILE,
        @SerializedName("challenge-mobile")
        CHALLENGE_MOBILE,
        @SerializedName("url")
        URL,
        @SerializedName("redirect")
        REDIRECT,
        @SerializedName("fingerprint")
        FINGERPRINT
    }
    
    @SerializedName("type")
    private CustomerActionType type;
    @SerializedName("value")
    private String value;

    public CustomerAction(CustomerActionType type, String value) {
        this.type = type;
        this.value = value;
    }

    public CustomerActionType getType() {
        return type;
    }

    public String getValue() {
        return value;
    }
}

File: sdk/src/main/kotlin/com/processout/sdk/api/model/response/CustomerAction.kt
package com.processout.sdk.api.model.response

import com.processout.sdk.core.util.findBy
import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
internal data class CustomerAction(
    @Json(name = "type")
    val rawType: String,
    val value: String
) {
    fun type() = Type::rawType.findBy(rawType) ?: Type.UNSUPPORTED

    @JsonClass(generateAdapter = false)
    enum class Type(val rawType: String) {
        FINGERPRINT_MOBILE("fingerprint-mobile"),
        CHALLENGE_MOBILE("challenge-mobile"),
        FINGERPRINT("fingerprint"),
        REDIRECT("redirect"),
        URL("url"),
        UNSUPPORTED(String())
    }
}
